### PR TITLE
Containerize integration tests

### PR DESCRIPTION
**Pull request by chris-crone**
**- What I did**
Containerized the integration tests so that they can be run against a remote daemon for end-to-end testing.

**- Why I did it**
There are many integration tests in _moby/moby_ that can be used elsewhere. Providing these as a container would facilitate their use.

**- How I did it**
A new build file can be found in the root of the repository: `DockerFile.e2e`. This describes a multistage build that first builds the tests and prepares their requirements before copying them into the resultant image.

Code was added to `integration-cli/environment` to protect existing elements (containers, networks, plugins, and volumes) so that the containerized tests can be run against a daemon with existing elements.

Some tests had to be reworked so that they work in environments where there are existing elements. Where possible, this was done by removing existing elements from outputs so that test logic would not need to be changed (see: `integration-cli/utils_tests.go:RemoveOutputForExistingElements`).

A couple of tests had to be reworked so that they use TLS when it is required by the remote daemon.

**- How to verify it**
1) Build the image: `docker build -t <image name> -f Docker.e2e .`
2) Run the tests against a remote daemon:
```bash
docker run --rm -it -e DOCKER_ENGINE_GOARCH=amd64 -e DOCKER_HOST=tcp://<daemon IP>:2376 -e DOCKER_TLS_VERIFY=1 -e DOCKER_CERT_PATH=/certs -v <path/to/certs>:/certs <image name>

Run integration/container tests
INFO: Testing against a remote daemon
=== RUN   TestCreateFailsWhenIdentifierDoesNotExist
=== RUN   TestCreateFailsWhenIdentifierDoesNotExist/image_and_tag
=== RUN   TestCreateFailsWhenIdentifierDoesNotExist/image_no_tag
=== RUN   TestCreateFailsWhenIdentifierDoesNotExist/digest
--- PASS: TestCreateFailsWhenIdentifierDoesNotExist (0.29s)
    --- PASS: TestCreateFailsWhenIdentifierDoesNotExist/image_no_tag (0.05s)
    --- PASS: TestCreateFailsWhenIdentifierDoesNotExist/image_and_tag (0.05s)
    --- PASS: TestCreateFailsWhenIdentifierDoesNotExist/digest (0.05s)
=== RUN   TestCreateWithInvalidEnv
=== RUN   TestCreateWithInvalidEnv/0
=== RUN   TestCreateWithInvalidEnv/1
=== RUN   TestCreateWithInvalidEnv/2
--- PASS: TestCreateWithInvalidEnv (0.29s)
    --- PASS: TestCreateWithInvalidEnv/2 (0.05s)
    --- PASS: TestCreateWithInvalidEnv/1 (0.06s)
    --- PASS: TestCreateWithInvalidEnv/0 (0.06s)
PASS
Run integration-cli DockerSuite tests
INFO: Testing against a remote daemon
=== RUN   Test
PASS: /go/src/github.com/docker/docker/integration-cli/docker_api_test.go:35: DockerSuite.TestAPIClientVersionOldNotSupported	0.028s
...
```

There is one test that fails, `DockerSuite.TestNetHostname`, but I believe that this PR will fix it: https://github.com/moby/moby/pull/34784.

**Note:** only tests in `integration/container` and `integration-cli` _{DockerSuite,DockerNetworkSuite}_ are currently run.

### CODE REVIEW

File: integration-cli/docker_cli_authz_plugin_v2_test.go [hunk 1]
```diff
@@ -72,7 +75,7 @@ func (s *DockerAuthzV2Suite) TestAuthZPluginAllowNonVolumeRequest(c *check.C) {
 
 	out, err = s.d.Cmd("ps")
 	c.Assert(err, check.IsNil)
-	c.Assert(assertContainerList(out, []string{id}), check.Equals, true)
+	c.Assert(assertContainerList(RemoveOutputForExistingElements(out, existingContainers), []string{id}), check.Equals, true)
```
Comment (dnephin) [inline integration-cli/docker_cli_authz_plugin_v2_test.go]: I don't really understand why this test should be looking at the container list. Is the intent to verify the container is actually running? I think the successful exit of `docker run`should already tell us.

I'd rather remove this check than compare against "existing elements".

cc @riyazdf: for context, we want to run these tests against different engines, some of which might have already running containers, so the output of `ps` is undefined.

File: integration-cli/docker_cli_authz_unix_test.go [hunk 2]
```diff
@@ -221,7 +223,7 @@ func (s *DockerAuthzSuite) TestAuthZPluginAllowRequest(c *check.C) {
 
 	out, err = s.d.Cmd("ps")
 	c.Assert(err, check.IsNil)
-	c.Assert(assertContainerList(out, []string{id}), check.Equals, true)
+	c.Assert(assertContainerList(RemoveOutputForExistingElements(out, existingContainers), []string{id}), check.Equals, true)
```
Comment (dnephin) [inline integration-cli/docker_cli_authz_unix_test.go]: (same here)

File: integration-cli/docker_cli_port_test.go [hunk 3]
```diff
@@ -148,9 +149,7 @@ func (s *DockerSuite) TestPortList(c *check.C) {
 
 	out, _ = dockerCmd(c, "port", ID)
 
-	err = assertPortList(c, out, []string{
-		"80/tcp -> 0.0.0.0:8000",
-		"80/udp -> 0.0.0.0:8000"})
+	err = assertPortRange(c, out, []int{8000, 8080}, []int{8000, 8080})
```
Comment (dnephin) [inline integration-cli/docker_cli_port_test.go]: Why is this change necessary, but the other `assertPortList` calls work fine?

File: integration-cli/utils_test.go [hunk 4]
```diff
@@ -112,3 +116,98 @@ func RunCommandPipelineWithOutput(cmds ...*exec.Cmd) (output string, err error)
...
+// a docker command.
+// This function takes an output string and returns a string.
+func RemoveOutputForExistingElements(output string, existing []string) string {
+	res := RemoveLinesForExistingElements(strings.Split(output, "\n"), existing)
+	return strings.Join(res, "\n")
+}
+
+// NewEnvClientWithVersion returns a docker client with a specified version.
+// See: github.com/docker/docker/client `NewEnvClient()`
+func NewEnvClientWithVersion(version string) (*client.Client, error) {
```
Comment (dnephin) [inline integration-cli/utils_test.go]: This should probably be in the `integration-cli/request` package

File: integration-cli/utils_test.go [hunk 5]
```diff
@@ -112,3 +116,98 @@ func RunCommandPipelineWithOutput(cmds ...*exec.Cmd) (output string, err error)
...
+	return v
+}
+
+// CheckKernelVersion checks if current kernel is newer than (or equal to)
+// the given version.
+func CheckKernelVersion(k, major, minor int) bool {
+	v := GetKernelVersion()
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: k, Major: major, Minor: minor}) < 0 {
+		return false
+	}
```
Comment (dnephin) [inline integration-cli/utils_test.go]: This `if` is unnecessary, just `return !kernel.CompareKernelVersion(...)`

Could this be fixed by moving `CheckKernelVersion()` out of the `_unx.go` into `kernel.go` ? It seems like this is the same implementation.

File: internal/test/environment/protect.go [hunk 6]
```diff
@@ -10,7 +10,63 @@ import (
...
+		networks:   map[string]struct{}{},
+		plugins:    map[string]struct{}{},
+		volumes:    map[string]struct{}{},
+	}
+}
+
+// ProtectAll protects the existing environment (containers, images, networks,
+// volumes, and, on Linux, plugins) from being cleaned up at the end of test
+// runs
+func ProtectAll(t testingT, testEnv *Execution) {
```
Comment (dnephin) [inline internal/test/environment/protect.go]: Calling `list all` for every object is going to add time to the test suite. It would be good to have some idea of how much extra time this adds.

**Review by dnephin — COMMENTED** [review]
Looking at this review I was reminded of something.

There are a bunch of tests, including most of the swarm tests, that don't actually run against the `DOCKER_HOST` swarm. They start one or more daemons, configure them as a swarm, and then run tests against it.

So exposing the docker socket will not actually work for the swarm tests.

File: integration-cli/docker_cli_port_test.go (hunk 3 above)
Comment (chris-crone) [reply integration-cli/docker_cli_port_test.go]: When you run this test multiple times, the TCP port used by the container increments (i.e.: first run its 8000, second run its 8001, etc.). I guess that this is because we pass a range (8000-8080) and the engine tries not to reuse recently used ports.

File: integration-cli/utils_test.go (hunk 5 above)
Comment (chris-crone) [reply integration-cli/utils_test.go]: I don't think so as `CheckKernelVersion(...)` fetches the kernel version and then compares it to the passed values. In `pkg/parsers/kernel`, this fetched value is the host kernel while my function fetches kernel information from the daemon.

We could move the `CheckKernelVersion(...)` out of `_unix.go` so that it could be used by other platforms but that's out of scope for this PR.

I'll make sure that I clean up the function return.

File: integration-cli/docker_cli_build_test.go [hunk 7]
```diff
@@ -388,6 +389,9 @@ func (s *DockerSuite) TestBuildCacheAdd(c *check.C) {
 }
 
 func (s *DockerSuite) TestBuildLastModified(c *check.C) {
+	// E2E: Requires built httpserver.
+	testRequires(c, NotE2E)
```
[the remark this answers is not in the document]
Comment (chris-crone) [reply integration-cli/docker_cli_build_test.go]: I've now worked around this: I will build the `httpserver` binary in the build step of the Dockerfile and then copy it to the test runner image.

File: Dockerfile.e2e [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    build-base \
+    curl \
+    lvm2-dev \
+    jq \
+    && rm -rf /var/cache/apk/*
+
+RUN mkdir -p /go/src/github.com/docker/docker/
+WORKDIR /go/src/github.com/docker/docker/
+
+COPY contrib contrib
```
[the remark this answers is not in the document]
Comment (chris-crone) [reply Dockerfile.e2e]: I've made this change.

Copying `contrib` here was just a build optimization. I fetch the frozen images before adding all the moby to ensure that the images are cached.

File: integration-cli/docker_api_images_test.go [hunk 9]
```diff
@@ -119,7 +119,8 @@ func (s *DockerSuite) TestAPIImagesHistory(c *check.C) {
 }
 
 func (s *DockerSuite) TestAPIImagesImportBadSrc(c *check.C) {
-	testRequires(c, Network)
+	// E2E: Test runs local HTTP server.
+	testRequires(c, Network, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_api_images_test.go]: Could this use `SameHostDaemon` instead of `NotE2E` ?

File: integration-cli/docker_cli_build_unix_test.go [hunk 10]
```diff
@@ -25,7 +25,8 @@ import (
 )
 
 func (s *DockerSuite) TestBuildResourceConstraintsAreUsed(c *check.C) {
-	testRequires(c, cpuCfsQuota)
+	// E2E: Test assumes no other containers running.
+	testRequires(c, cpuCfsQuota, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_build_unix_test.go]: I think we should fix this test instead of skipping it. There's no reason a build test should need a clean environment.

I think this can be fixed by adding a `--label` to the build command and then filtering by that label, but there may be an easier way.

File: integration-cli/docker_cli_network_unix_test.go [hunk 11]
```diff
@@ -289,7 +289,8 @@ func (s *DockerNetworkSuite) TestDockerNetworkLsDefault(c *check.C) {
 }
 
 func (s *DockerSuite) TestNetworkLsFormat(c *check.C) {
-	testRequires(c, DaemonIsLinux)
+	// E2E: Test assumes only default networks.
+	testRequires(c, DaemonIsLinux, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_network_unix_test.go]: This test is only testing CLI behaviour, please delete it instead of skipping it (see #32866).

We already test the formatter in `docker/cli`

File: integration-cli/docker_cli_network_unix_test.go [hunk 12]
```diff
@@ -300,7 +301,8 @@ func (s *DockerSuite) TestNetworkLsFormat(c *check.C) {
 }
 
 func (s *DockerSuite) TestNetworkLsFormatDefaultFormat(c *check.C) {
-	testRequires(c, DaemonIsLinux)
+	// E2E: Test assumes only default networks.
+	testRequires(c, DaemonIsLinux, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_network_unix_test.go]: please delete this test, only tests the CLI

File: integration-cli/docker_cli_network_unix_test.go [hunk 13]
```diff
@@ -351,6 +353,8 @@ func (s *DockerNetworkSuite) TestDockerNetworkRmPredefined(c *check.C) {
 }
 
 func (s *DockerNetworkSuite) TestDockerNetworkLsFilter(c *check.C) {
+	// E2E: Test assumes clean network environment.
+	testRequires(c, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_network_unix_test.go]: This test is unfortunately not written very well because it actually tests many cases in one.

Instead of skipping the entire function could we put an `if CleanEnvironment()` around the cases that don't work?

`CleanEnvironment()` can be similar to `NotE2E` but would use a different env variable name.

File: integration-cli/requirements_test.go [hunk 14]
```diff
@@ -36,6 +36,14 @@ func DaemonIsLinux() bool {
 	return testEnv.DaemonInfo.OSType == "linux"
 }
 
+func E2E() bool {
+	return os.Getenv("DOCKER_E2E") != ""
+}
```
Comment (dnephin) [inline integration-cli/requirements_test.go]: Commented on all the uses of this, I think this should be replaced with something that describes the requirement (a prestine/clean engine with no other resources) instead of the environment in which it runs (E2E). That will make it a lot easier to maintain because the requirement is more obvious from the name.

I would suggest something like `TEST_IS_PRESTINE_ENGINE` or something like that for the environment variable.

File: integration-cli/docker_cli_ps_test.go [hunk 15]
```diff
@@ -509,8 +519,9 @@ func (s *DockerSuite) TestPsRightTagName(c *check.C) {
 }
 
 func (s *DockerSuite) TestPsLinkedWithNoTrunc(c *check.C) {
+	// E2E: Test assumes no other containers.
 	// Problematic on Windows as it doesn't support links as of Jan 2016
-	testRequires(c, DaemonIsLinux)
+	testRequires(c, DaemonIsLinux, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_ps_test.go]: This test can be fixed by using a label and filter, or by just checking that the one expected lines are in the output and ignoring the rest.

File: integration-cli/docker_cli_ps_test.go [hunk 16]
```diff
@@ -604,19 +618,24 @@ func (s *DockerSuite) TestPsFormatMultiNames(c *check.C) {
...
+	lines = RemoveLinesForExistingElements(lines, existingContainers)
 	expected := []string{"test2 test2", "test1 test1"}
 	var names []string
 	names = append(names, lines...)
 	c.Assert(expected, checker.DeepEquals, names, check.Commentf("Expected array with names displayed twice: %v, got: %v", expected, names))
 }
 
 func (s *DockerSuite) TestPsFormatHeaders(c *check.C) {
+	// E2E: Test assumes no other containers.
+	testRequires(c, NotE2E)
```
Comment (dnephin) [inline integration-cli/docker_cli_ps_test.go]: Please delete this test, it's only testing the CLI.

File: hack/test/e2e-run.sh [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+set -e
+
+export DOCKER_ENGINE_GOARCH=${DOCKER_ENGINE_GOARCH:-amd64}
+
+echo "Ensure emptyfs image is loaded"
+bash /ensure-emptyfs.sh
+
+echo "Run integration/container tests"
+cd /tests/integration/container
+./test.main -test.v
```
Comment (dnephin) [inline hack/test/e2e-run.sh]: This script should probably make use of https://github.com/moby/moby/blob/master/hack/make/.integration-test-helpers#L25-L34 so that as we make changes to the moby/moby build this script wont break.

At the very least it should iterate over all the packages in that directory.

File: hack/test/e2e-run.sh [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+echo "Run integration-cli tests (DockerSuite, DockerNetworkSuite)"
+cd /tests/integration-cli
+./test.main -test.v -check.v -check.f "DockerSuite|DockerNetworkSuite"
```
Comment (dnephin) [inline hack/test/e2e-run.sh]: Why only these 2 suites? Do the other suites all expect to start new daemons? 

A comment about why it only runs 2 suites would be great.

File: integration-cli/cli/build/fakestorage/fixtures.go [hunk 19]
```diff
@@ -39,27 +39,34 @@ func ensureHTTPServerImage(t testingT) {
...
+	if _, err = os.Stat("../contrib/httpserver/httpserver"); os.IsNotExist(err) {
+		goCmd, lookErr := exec.LookPath("go")
+		if lookErr != nil {
+			t.Fatalf("could not build http server: %v", lookErr)
+		}
 
-	cpCmd, lookErr := exec.LookPath("cp")
-	if lookErr != nil {
-		t.Fatalf("could not build http server: %v", lookErr)
+		cmd := exec.Command(goCmd, "build", "-o", filepath.Join(tmp, "httpserver"), "github.com/docker/docker/contrib/httpserver")
```
Comment (dnephin) [inline integration-cli/cli/build/fakestorage/fixtures.go]: I like your solution of just building this during setup. We could remove this `go build` and do it from https://github.com/moby/moby/blob/master/hack/make/test-integration#L10 instead. I guess it's not critical for this PR , and we can fix it alter.

File: integration-cli/utils_unix_test.go [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+
+// GetKernelVersion gets the current kernel version.
+func GetKernelVersion() *kernel.VersionInfo {
+	v, _ := kernel.ParseRelease(testEnv.DaemonInfo.KernelVersion)
+	return v
+}
+
+// CheckKernelVersion checks if current kernel is newer than (or equal to)
+// the given version.
+func CheckKernelVersion(k, major, minor int) bool {
```
Comment (dnephin) [inline integration-cli/utils_unix_test.go]: Since this is only used in a single place please move these 2 functions to immediately after the test which uses them.

File: internal/test/environment/protect.go (hunk 6 above)
Comment (dnephin) [reply internal/test/environment/protect.go]: Looking at a few of the janky builds it seems the job run time is about the same, so that's reassuring.

File: Dockerfile.e2e [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    iptables \
+    tar \
+    xz \
+    && rm -rf /var/cache/apk/*
+
+# Add an unprivileged user to be used for tests which need it
+RUN addgroup docker && adduser -D -G docker unprivilegeduser -s /bin/ash
+
+COPY contrib/httpserver/Dockerfile /tests/contrib/httpserver/Dockerfile
+COPY contrib/syscall-test /tests/contrib/syscall-test
```
[the remark this answers is not in the document]
Comment (chris-crone) [reply Dockerfile.e2e]: I believe there will be another couple that we need for Windows. Notably, we need to build the faux busybox image: https://raw.githubusercontent.com/jhowardmsft/busybox/master/Dockerfile

File: integration-cli/docker_cli_network_unix_test.go (hunk 13 above)
Comment (chris-crone) [reply integration-cli/docker_cli_network_unix_test.go]: With the other changes you suggested, I've removed the `E2E` requirement. For this test, I added an `OnlyDefaultNetworks` requirement as that's more precise.

File: integration-cli/requirements_test.go (hunk 14 above)
Comment (chris-crone) [reply integration-cli/requirements_test.go]: I don't think there are any tests remaining that require this so will not add it now.

File: hack/test/e2e-run.sh (hunk 17 above)
Comment (chris-crone) [reply hack/test/e2e-run.sh]: I didn't use `.integration-test-helpers` because it depends on lots of other scripts. I will copy the functions used there though as they're cleaner than what I did.

File: hack/test/e2e-run.sh (hunk 18 above)
Comment (chris-crone) [reply hack/test/e2e-run.sh]: It's simply that those are the suites I've made work for E2E. Combined they represent ~1200 of the ~1600 total tests. There may be some other suites that we can run but some we will never be able to run in this way, namely `DockerDaemonSuite`, `DockerSwarmSuite`, and `DockerPluginSuite`.

I am continuing work to bring up other suites but will make that more clear in the script.

**Comment by chris-crone** [comment]
The only `integration-cli` test suites that currently work against a remote socket are `DockerSuite` and `DockerNetworkSuite`. This represents the majority of the tests though (~1000).

There might be a couple of others that can easily be made to work but I suggest that we add those as a separate PR.

**Review by dnephin — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐯 🎉

**Comment by thaJeztah** [comment]
:tada:
